internal/indexer/processors/effects.go

```diff
@@ -0,0 +1,255 @@
+package processors
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/stellar/go/ingest"
+	effects "github.com/stellar/go/processors/effects"
+	operation "github.com/stellar/go/processors/operation"
+	"github.com/stellar/go/toid"
+	"github.com/stellar/go/xdr"
+
+	"github.com/stellar/wallet-backend/internal/indexer/types"
+)
+
+var (
+	thresholdToReasonMap = map[string]types.StateChangeReason{
+		"low_threshold":  types.StateChangeReasonLow,
+		"med_threshold":  types.StateChangeReasonMedium,
+		"high_threshold": types.StateChangeReasonHigh,
+	}
+	signerEffectToReasonMap = map[int32]types.StateChangeReason{
+		int32(effects.EffectSignerCreated): types.StateChangeReasonAdd,
+		int32(effects.EffectSignerRemoved): types.StateChangeReasonRemove,
+		int32(effects.EffectSignerUpdated): types.StateChangeReasonUpdate,
+	}
+	accountFlags = []string{
+		"auth_required_flag",
+		"auth_revocable_flag",
+		"auth_immutable_flag",
+		"auth_clawback_enabled_flag",
+	}
+	trustlineFlags = []string{
+		"authorized_flag",
+		"authorized_to_maintain_liabilites",
+		"clawback_enabled_flag",
+	}
+)
+
+type EffectsProcessor struct {
+	networkPassphrase string
+}
+
+func NewEffectsProcessor(networkPassphrase string) *EffectsProcessor {
+	return &EffectsProcessor{
+		networkPassphrase: networkPassphrase,
+	}
+}
+
+func (p *EffectsProcessor) ProcessTransaction(ctx context.Context, tx ingest.LedgerTransaction, op xdr.Operation, opIdx uint32) ([]types.StateChange, error) {
+	ledgerCloseTime := tx.Ledger.LedgerCloseTime()
+	ledgerNumber := tx.Ledger.LedgerSequence()
+	opID := toid.New(int32(ledgerNumber), int32(tx.Index), int32(opIdx)).ToInt64()
+	txID := toid.New(int32(ledgerNumber), int32(tx.Index), 0).ToInt64()
+
+	opWrapper := operation.TransactionOperationWrapper{
+		Index:          opIdx,
+		LedgerSequence: tx.Ledger.LedgerSequence(),
+		LedgerClosed:   time.Unix(tx.Ledger.LedgerCloseTime(), 0),
+		Transaction:    tx,
+		Operation:      op,
+		Network:        p.networkPassphrase,
+	}
+
+	effectOutputs, err := effects.Effects(&opWrapper)
+	if err != nil {
+		return nil, fmt.Errorf("processing effects: %w", err)
+	}
+
+	stateChanges := make([]types.StateChange, 0)
+	masterBuilder := NewStateChangeBuilder(ledgerNumber, ledgerCloseTime, txID).WithOperationID(opID)
+	for _, effect := range effectOutputs {
+		changeBuilder := masterBuilder.Clone().WithAccount(effect.Address)
+
+		effectType := effects.EffectType(effect.Type)
+		//exhaustive:ignore
+		switch effectType {
+		case effects.EffectSignerCreated, effects.EffectSignerRemoved, effects.EffectSignerUpdated:
+			signerWeight := effect.Details["weight"]
+			signerPublicKey := effect.Details["public_key"]
+			stateChanges = append(stateChanges, changeBuilder.
+				WithCategory(types.StateChangeCategorySigner).
+				WithReason(signerEffectToReasonMap[effect.Type]).
+				WithSigner(signerPublicKey.(string), signerWeight).
+				Build())
```

Comment:
Looking at the design doc, we were going to have the old and new weights for a signer when it was updated, but I don't see that in the implementation. Also when a signer is removed, should we index its weight? The design doc omits it, but its present here.

Reply:
@JakeUrban Where did you read about old and new weights for signer? We have old and new threshold values but not for weights 🤔

Although I am not arguing against storing the old weights. I think we should do it to be uniform with the thresholds - https://github.com/stellar/wallet-backend/pull/224/commits/3d43680d6d283c70ecc6dd857e83712066e18c9d